src/Workspaces/Core/Portable/CodeActions/Operations/ApplyChangesOperation.cs

```diff
@@ -38,12 +42,113 @@ public ApplyChangesOperation(Solution changedSolution)
+                solutionChanges.GetAddedAnalyzerReferences().Count() > 0 ||
+                solutionChanges.GetRemovedProjects().Count() > 0 ||
+                solutionChanges.GetRemovedAnalyzerReferences().Count() > 0)
+            {
+                return false;
+            }
+
+            // Take the actual current solution the workspace is pointing to and fork it with just the text changes the
+            // code action wanted to make.  Then apply that fork back into the workspace.
+            var forkedSolution = currentSolution;
+
+            foreach (var changedProject in solutionChanges.GetProjectChanges())
+            {
+                // We only support text changes.  If we see any other changes to this project, bail out immediately.
+                if (changedProject.GetAddedAdditionalDocuments().Count() > 0 ||
+                    changedProject.GetAddedAnalyzerConfigDocuments().Count() > 0 ||
+                    changedProject.GetAddedAnalyzerReferences().Count() > 0 ||
+                    changedProject.GetAddedDocuments().Count() > 0 ||
+                    changedProject.GetAddedMetadataReferences().Count() > 0 ||
+                    changedProject.GetAddedProjectReferences().Count() > 0 ||
+                    changedProject.GetRemovedAdditionalDocuments().Count() > 0 ||
+                    changedProject.GetRemovedAnalyzerConfigDocuments().Count() > 0 ||
+                    changedProject.GetRemovedAnalyzerReferences().Count() > 0 ||
+                    changedProject.GetRemovedDocuments().Count() > 0 ||
+                    changedProject.GetRemovedMetadataReferences().Count() > 0 ||
+                    changedProject.GetRemovedProjectReferences().Count() > 0)
+                {
+                    return false;
+                }
+
+                if (!TryForkTextChanges(changedProject, changedProject.GetChangedDocuments(), static (s, i) => s.GetRequiredDocument(i), static (s, i, t) => s.WithDocumentText(i, t)) ||
+                    !TryForkTextChanges(changedProject, changedProject.GetChangedAdditionalDocuments(), static (s, i) => s.GetRequiredAdditionalDocument(i), static (s, i, t) => s.WithAdditionalDocumentText(i, t)) ||
+                    !TryForkTextChanges(changedProject, changedProject.GetChangedAnalyzerConfigDocuments(), static (s, i) => s.GetRequiredAnalyzerConfigDocument(i), static (s, i, t) => s.WithAnalyzerConfigDocumentText(i, t)))
+                {
+                    return false;
+                }
+            }
+
+            return workspace.TryApplyChanges(forkedSolution, progressTracker);
+
+            bool TryForkTextChanges(
+                ProjectChanges changedProject,
+                IEnumerable<DocumentId> changedDocuments,
+                Func<Solution, DocumentId, TextDocument> getDocument,
+                Func<Solution, DocumentId, SourceText, Solution> withDocumentText)
+            {
+                var sawChangedDocument = false;
```

Comment:
should this return true if changedDocuments is empty?  It seems like the method should only return false if we hit any of the false conditions in the loop

Reply:
yeah, this was bogus.